server/text_generation_server/models/vlm_causal_lm.py

```diff
@@ -190,6 +232,30 @@ def batch_tokenized_inputs(
                         images.append(image)
                     else:
                         images.append([image])
+                elif chunk_type == "video":
+                    if config.model_type == "qwen2_vl":
+                        video_frame_buf = np.frombuffer(
+                            chunk.video.data, dtype=np.uint8
+                        )
+                        num_bytes = len(video_frame_buf)
+                        bytes_per_frame = num_bytes // chunk.video.frames
+
+                        # iterate over with a stride the size of a frame
+                        frames = []
+                        for i in range(chunk.video.frames):
+                            frame = video_frame_buf[
+                                i * bytes_per_frame : (i + 1) * bytes_per_frame
+                            ]
+                            frame = frame.reshape(
+                                chunk.video.height, chunk.video.width, 3
+                            )
+                            frames.append(frame)
+
+                        video_frame_buf = np.stack(frames)
+                        frame_nchw_tensor = torch.from_numpy(video_frame_buf).permute(
+                            0, 3, 1, 2
```

Comment:
This can and should be done without reallocating once.

Reply:
good point, the latest commit updates this to avoid the loop and prefer a vectorized method with numpy and torch.